### PR TITLE
[feat] experimental: Add spectrain support

### PR DESCRIPTION
**Pull request by sidgoyal78**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Adds support for training models using spectrain based asynchronous pipelining. Reference: https://arxiv.org/pdf/1809.02839.pdf
 
## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: benchmarks/benchmark_dataset.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import torch
```
Comment (msbaines) [inline benchmarks/benchmark_dataset.py:1]: Please add a copyright notice.

**Review by msbaines — APPROVED** [review]

File: benchmarks/benchmark_dataset.py (hunk 1 above)
Comment (sidgoyal78) [inline benchmarks/benchmark_dataset.py:1]: @msbaines : should I put this file and the experimental_ampnet.py inside an benchmarks/experimental/ folder.

Looks like this file (benchmark_dataset.py) was removed earlier.

Comment (anj-s) [reply benchmarks/benchmark_dataset.py:1]: drive by comment: Mentioned in another conversation, we refactored fairscale benchmarks and have common dataset loaders. We should try and use that unless it doesn't work for your use case.

File: benchmarks/experimental_ampnet.py [hunk 2]
```diff
@@ -157,6 +157,108 @@ def step(self, closure=None):
...
+        for more details.
+
+        Args:
+            params (iterable): iterable of parameters to optimize or dicts defining
+                parameter groups
+            lr (float): learning rate (required)
+    """
+
+    def __init__(self, params, lr=0.01, momentum=0.9):
+        print("called")
```
Comment (anj-s) [inline benchmarks/experimental_ampnet.py:173]: remove?

File: benchmarks/experimental_ampnet.py [hunk 3]
```diff
@@ -518,6 +613,7 @@ def bench_mpi(args):
 parser.add_argument("--max-batch", type=int, default=4, help="Max number of batches")
 parser.add_argument("--socket-name", type=str, default=None, help="socket ifname for gloo/tp")
 parser.add_argument("--num-decoder-layers", type=int, default=10, help="Number of decoder layers in the model")
+parser.add_argument("--spectrain", action="store_true", default=False, help="Use spectrain based weight prediction")
```
Comment (anj-s) [inline benchmarks/experimental_ampnet.py:616]: Do we want to enable these benchmarks in circleCI?

File: benchmarks/experimental_ampnet.py [hunk 4]
```diff
@@ -157,6 +157,108 @@ def step(self, closure=None):
...
+class SpectrainSGDMomentum(Optimizer):
+    r"""
+        Implements a SGD with momentum optimizer with Spectrain based weight
+        prediction. Please refer to the spectrain paper: https://arxiv.org/pdf/1809.02839.pdf
+        for more details.
+
+        Args:
+            params (iterable): iterable of parameters to optimize or dicts defining
+                parameter groups
+            lr (float): learning rate (required)
```
[the remark this answers is not in the document]
Comment (sidgoyal78) [reply benchmarks/experimental_ampnet.py:169]: Sounds good. Fixed MySGD constructor as well :)

File: benchmarks/experimental_ampnet.py (hunk 2 above)
Comment (sidgoyal78) [reply benchmarks/experimental_ampnet.py:173]: Done

File: benchmarks/experimental_ampnet.py [hunk 5]
```diff
@@ -157,6 +157,108 @@ def step(self, closure=None):
...
+            for group in self.param_groups:
+                multiplier = group["lr"] * (model_index // 2)
+                for p in group["params"]:
+                    param_state = self.state[p]
+                    p.data.sub_(param_state["momentum_buffer"].data, alpha=multiplier)
+
+    def step(self, weight_prediction=True, closure=None):
+        """ Performs a single optimization step.
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
```
[the remark this answers is not in the document]
Comment (sidgoyal78) [reply benchmarks/experimental_ampnet.py:235]: fixed

File: benchmarks/experimental_ampnet.py (hunk 3 above)
Comment (sidgoyal78) [reply benchmarks/experimental_ampnet.py:616]: I think we can do that later.

File: benchmarks/benchmark_dataset.py (hunk 1 above)
Comment (sidgoyal78) [reply benchmarks/benchmark_dataset.py:1]: Yeah, that's a great suggestion. I plan to refactor this, but I will first make a PR with xpipe (which depends on the dataloader from this script). Once we merge that, I will plan to refactor.

**Comment by sidgoyal78** [comment]
@anj-s and @msbaines : Thanks for reviewing the PR. I addressed most of your comments, and would be great if you could take a final look.

File: benchmarks/benchmark_dataset.py (hunk 1 above)
Comment (sidgoyal78) [reply benchmarks/benchmark_dataset.py]: Done.

File: benchmarks/experimental/benchmark_dataset.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+
+import torch
+from torch.utils.data import Dataset
+
+
```
Comment (anj-s) [inline benchmarks/experimental/benchmark_dataset.py:6]: Can we add a TODO that this file will be removed?

File: benchmarks/experimental/benchmark_dataset.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
```
Comment (anj-s) [inline benchmarks/experimental/benchmark_dataset.py:1]: Missing license section?

File: benchmarks/experimental/experimental_ampnet.py [hunk 8]
```diff
@@ -157,6 +157,109 @@ def step(self, closure=None):
...
+            # 2. predict new weights and modify
+
+            self.modify_reference_params_using_current_params()
+            for group in self.param_groups:
+                multiplier = group["lr"] * (model_index // 2 + num_gpus - model_index - 1)
+                for p in group["params"]:
+                    param_state = self.state[p]
+                    p.data.sub_(param_state["momentum_buffer"].data, alpha=multiplier)
+
+        if not forward:
```
Comment (anj-s) [inline benchmarks/experimental/experimental_ampnet.py:220]: nit: why not an "else" clause?

File: benchmarks/experimental/experimental_ampnet.py [hunk 9]
```diff
@@ -157,6 +157,109 @@ def step(self, closure=None):
...
+            self.modify_current_params_using_reference_params()
+            for group in self.param_groups:
+                multiplier = group["lr"] * (model_index // 2)
+                for p in group["params"]:
+                    param_state = self.state[p]
+                    p.data.sub_(param_state["momentum_buffer"].data, alpha=multiplier)
+
+    def step(self, weight_prediction=True, closure=None):
+        """ Performs a single optimization step.
+        Arguments:
```
Comment (anj-s) [inline benchmarks/experimental/experimental_ampnet.py:234]: s/Arguments/Args?

**Review by anj-s — APPROVED** [review]

**Comment by anj-s** [comment]
> @anj-s and @msbaines : Thanks for reviewing the PR. I addressed most of your comments, and would be great if you could take a final look.

thank you @sidgoyal78 for the PR and making changes! Another thing to mention is that the model can also be reused similar to benchmarks/pipe.py when you end up refactoring.

File: benchmarks/experimental/experimental_ampnet.py (hunk 9 above)
Comment (sidgoyal78) [reply benchmarks/experimental/experimental_ampnet.py]: fixed

File: benchmarks/experimental/experimental_ampnet.py (hunk 8 above)
Comment (sidgoyal78) [reply benchmarks/experimental/experimental_ampnet.py]: fixed, thanks for pointing.

File: benchmarks/experimental/benchmark_dataset.py (hunk 6 above)
Comment (sidgoyal78) [reply benchmarks/experimental/benchmark_dataset.py:8]: Done

File: benchmarks/experimental/benchmark_dataset.py (hunk 7 above)
Comment (sidgoyal78) [reply benchmarks/experimental/benchmark_dataset.py:1]: I think there's a discrepancy across many scripts. I noticed that this is the comment that is present in most of the other scripts (see, benchmarks/pipe.py, etc). However, there's an extra license section in benchmarks/experimental/offload.py 

Let me make an issue and we can address it separately.

**Pull request merged**